File: src/state.ts
import { mkdir, readFile, rename, writeFile } from "node:fs/promises";
import path from "node:path";

export type DocumentStatus = "submitted" | "completed" | "failed" | "processing" | "unknown" | string;

export type DocumentRecord = {
  docId: string;
  originalName: string;
  uploadedPath: string;
  submittedPath: string;
  status: DocumentStatus;
  createdAt: string;
};

type Manifest = {
  documents: DocumentRecord[];
};

export function newDocumentRecord(
  docId: string,
  originalName: string,
  uploadedPath: string,
  submittedPath: string
): DocumentRecord {
  return {
    docId,
    originalName,
    uploadedPath,
    submittedPath,
    status: "submitted",
    createdAt: new Date().toISOString()
  };
}

export class ManifestStore {
  constructor(private readonly manifestPath: string) {}

  async list(): Promise<DocumentRecord[]> {
    try {
      const raw = await readFile(this.manifestPath, "utf8");
      const parsed = JSON.parse(raw) as Manifest;
      return parsed.documents ?? [];
    } catch (error) {
      if ((error as NodeJS.ErrnoException).code === "ENOENT") {
        return [];
      }
      throw error;
    }
  }

  async get(docId: string): Promise<DocumentRecord | undefined> {
    return (await this.list()).find((record) => record.docId === docId);
  }

  async upsert(record: DocumentRecord): Promise<void> {
    const records = (await this.list()).filter((item) => item.docId !== record.docId);
    records.push(record);
    await this.write(records);
  }

  async updateStatus(docId: string, status: DocumentStatus): Promise<void> {
    const records = await this.list();
    const record = records.find((item) => item.docId === docId);
    if (record) {
      record.status = status;
      await this.write(records);
    }
  }

  async completedIds(): Promise<string[]> {
    return (await this.list()).filter((record) => record.status === "completed").map((record) => record.docId);
  }

  private async write(records: DocumentRecord[]): Promise<void> {
    await mkdir(path.dirname(this.manifestPath), { recursive: true });
    const tmpPath = `${this.manifestPath}.tmp`;
    await writeFile(tmpPath, JSON.stringify({ documents: records }, null, 2), "utf8");
    await rename(tmpPath, this.manifestPath);
  }
}

File: src/chatbot_pipeline/state.py
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
from typing import Iterable


@dataclass
class DocumentRecord:
    doc_id: str
    original_name: str
    uploaded_path: str
    submitted_path: str
    status: str
    created_at: str

    @classmethod
    def new(cls, doc_id: str, original_name: str, uploaded_path: Path, submitted_path: Path) -> "DocumentRecord":
        return cls(
            doc_id=doc_id,
            original_name=original_name,
            uploaded_path=str(uploaded_path),
            submitted_path=str(submitted_path),
            status="submitted",
            created_at=datetime.now(timezone.utc).isoformat(),
        )


class ManifestStore:
    def __init__(self, path: Path) -> None:
        self.path = path
        self.path.parent.mkdir(parents=True, exist_ok=True)

    def list(self) -> list[DocumentRecord]:
        if not self.path.exists():
            return []
        data = json.loads(self.path.read_text(encoding="utf-8"))
        return [DocumentRecord(**item) for item in data.get("documents", [])]

    def get(self, doc_id: str) -> DocumentRecord | None:
        return next((record for record in self.list() if record.doc_id == doc_id), None)

    def upsert(self, record: DocumentRecord) -> None:
        records = [item for item in self.list() if item.doc_id != record.doc_id]
        records.append(record)
        self._write(records)

    def update_status(self, doc_id: str, status: str) -> None:
        records = self.list()
        for record in records:
            if record.doc_id == doc_id:
                record.status = status
                self._write(records)
                return

    def completed_ids(self) -> list[str]:
        return [record.doc_id for record in self.list() if record.status == "completed"]

    def _write(self, records: Iterable[DocumentRecord]) -> None:
        payload = {"documents": [asdict(record) for record in records]}
        tmp_path = self.path.with_suffix(".tmp")
        tmp_path.write_text(json.dumps(payload, indent=2), encoding="utf-8")
        tmp_path.replace(self.path)
